### ISSUE TITLE
Panama vs. Pennsylvania

### ISSUE THREAD

**Issue by missinglink** [openvenues/libpostal#344]
Hi!

I was checking out the awesome libpostal, and saw something that could be improved.

---
#### My country is
USA

---
#### Here's how I'm using libpostal
Pelias

---
#### Here's what I did
```
./address_parser
```

---
#### Here's what I got
```
> 128 Ashley Way, PA

Result:

{
  "house_number": "128",
  "road": "ashley way",
  "country": "pa"
}
```

---
#### Here's what I was expecting
```
128 Ashley Way, Plymouth Meeting, PA 19462, USA
```

---
#### For parsing issues, please check all that apply

- [x] Does the input address exist in [OpenStreetMap](https://openstreetmap.org)?
- [x] Do all the toponyms exist in OSM (city, state, region names, etc.)?
- [x] If the address uses a rare/uncommon format, does changing the order of the fields yield the correct result?
- [x] If the address does not contain city, region, etc., does adding those fields to the input improve the result?
- [x] If the address contains apartment/floor/sub-building information or uncommon formatting, does removing that help? Is there any minimum form of the address that gets the right parse?

---
#### Here's what I think could be improved

Hey Al,

I noticed this case today which is being classified as `Panama` when it would really be `Pennsylvania`.

osm: https://nominatim.openstreetmap.org/details.php?place_id=208713070

**Comment by albarrentine** [comment]
Related to #254. The 1.0 training data was built with the requirement that either `city` or an equivalent like {`suburb`, `city_district`} be listed on the example where available.

The goal was to avoid forcing the parser model to learn from conflicting information. In 0.3, the process was to reverse geocode an address point to all containing boundaries and then drop out various boundaries at random. This creates many different types of training examples, but can also create some pathological problems in the training set e.g. where the "New York" in "123 4th St, New York" means `city` in one instance and `state` in the next. Around the world there were many cases where an urban district might have the the same boundary and same name as a city, so could create many examples with conflicting labels, when in practice a city's name should always mean the city by default, and could optionally also have the district following it. The district would never stand alone.

On the other hand, `country` is allowed on its own in some places in the world, including Panama, so a transition out of a `road` tag into "PA" would only ever really exist in Panama in the current training data. There effectively are no training examples like "123 4th St, PA" for Pennsylvania. It would always be "123 4th St, Philadelphia, PA." But there might be a "Calle 123, 4 PA" in Panama. In other unambiguous cases like listing "QC" without a city (where there's only one interpretation), it will correctly interpret QC as `state` but may have a strong weight on the previous token being a `city`, regardless of whether it is or whether the combo makes sense.

If the decision of when to drop the city name is arbitrary and random, as it was in v0.3, we'd end up with many of the ambiguous cases mentioned. Instead we said: `city` must be included as a component 100% of the times where a city exists for a given address point (either on the record itself or through reverse geocoding) and we're including place names in the example (i.e. we're not just creating a simple street address like "123 4th St"), and if there are any exceptions to that we'd list them out as we go.

That's a bit overzealous, and may lead to some issues where the parser really wants there to be a city name based on the structure of the address. To my knowledge, it's a relatively rare edge case for a human to write other place names like `state` without a city preceding (that type of PA example I've only ever really seen in Pelias tests, but not in the wild). It's a bit more likely in the case of machine-generated addresses where there are some NULL values, as can happen in OpenAddresses, scraping from dynamic websites, etc.

What I've been considering for the next version is, rather than worry about the nomenclature for tagging places everywhere in the world or tracking down all the formatting exceptions, we'd just parse every toponym as a single tag like `place`, delimiting each phrase separately, and that way we can import a wide variety of place hierarchies (WoF, GeoNames, etc.) without needing to map the exceptions for arrondissements in France in every data set, or worry as much about including every possible format that may be encountered in the wild. It would also cut down on the number of tags and make the model smaller.

From there, it's possible to geocode all the tagged places, either to a fixed hierarchy like WoF or something user-supplied. That's probably more consistent with what people usually want to do with place parsing, and it has benefits for deduping as well (could match addresses/venues where one data set lists only the county and the other lists the city or neighborhood).
